exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -0,0 +1,230 @@
+ILQRSolver::~ILQRSolver() = default;
+
+void ILQRSolver::SpecifyProblem(PlanningProblemPtr pointer)
+{
+    if (pointer->type() != "exotica::DynamicTimeIndexedShootingProblem")
+    {
+        ThrowNamed("This ILQRSolver can't solve problem of type '" << pointer->type() << "'!");
+    }
+    MotionSolver::SpecifyProblem(pointer);
+    prob_ = std::static_pointer_cast<DynamicTimeIndexedShootingProblem>(pointer);
+    dynamics_solver_ = prob_->GetScene()->GetDynamicsSolver();
+    if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "initialized");
+}
+
+void ILQRSolver::BackwardPass()
+{
+    constexpr double min_clamp_ = -1e10;
+    constexpr double max_clamp_ = 1e10;
+    int T = prob_->get_T();
+
+    const Eigen::MatrixXd Qf = prob_->get_Qf(), R = prob_->get_R();
+    const Eigen::MatrixXd X_star = prob_->get_X_star();
+
+    // eq. 18
+    vk_gains_[T - 1] = Qf * dynamics_solver_->StateDelta(prob_->get_X(T - 1), X_star.col(T - 1));
+    vk_gains_[T - 1] = vk_gains_[T - 1].unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+        return std::min(std::max(x, min_clamp_), max_clamp_);
+    });
+
+    Eigen::MatrixXd Sk = Qf;
+    Sk = Sk.unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+        return std::min(std::max(x, min_clamp_), max_clamp_);
+    });
+
+    for (int i = T - 2; i > 0; i--)
+    {
+        Eigen::VectorXd x = prob_->get_X(i), u = prob_->get_U(i);
+        Eigen::MatrixXd Ak = dynamics_solver_->fx(x, u), Bk = dynamics_solver_->fu(x, u),
+                        Q = prob_->get_Q(i);
+
+        // eq. 13-16
+        Ak = Ak * dynamics_solver_->get_dt() + Eigen::MatrixXd::Identity(Ak.rows(), Ak.cols());
+        Bk = Bk * dynamics_solver_->get_dt();
+        // this inverse is common for all factors
+        const Eigen::MatrixXd _inv =
+            (Eigen::MatrixXd::Identity(R.rows(), R.cols()) * 1e-5 + R + Bk.transpose() * Sk * Bk).inverse();
+
+        Kv_gains_[i] = _inv * Bk.transpose();
+        K_gains_[i] = _inv * Bk.transpose() * Sk * Ak;
+        Ku_gains_[i] = _inv * R;
+        Sk = Ak.transpose() * Sk * (Ak - Bk * K_gains_[i]) + Q;
+
+        vk_gains_[i] = ((Ak - Bk * K_gains_[i]).transpose() * vk_gains_[i + 1]) -
+                       (K_gains_[i].transpose() * R * u) + (Q * x);
+
+        // fix for large values
+        Sk = Sk.unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+            return std::min(std::max(x, min_clamp_), max_clamp_);
+        });
+        vk_gains_[i] = vk_gains_[i].unaryExpr([min_clamp_, max_clamp_](double x) -> double {
+            return std::min(std::max(x, min_clamp_), max_clamp_);
+        });
+    }
+}
+
+double ILQRSolver::ForwardPass(double alpha, Eigen::MatrixXdRef ref_trajectory)
+{
+    double cost = 0;
+    int T = prob_->get_T();
+    const Eigen::VectorXd control_limits = dynamics_solver_->get_control_limits();
+
+    DynamicsSolverPtr dynamics_solver_ = prob_->GetScene()->GetDynamicsSolver();
+    for (int i = 0; i < T - 1; ++i)
```

Comment:
nit: `t` instead of `i` would be more appropriate - ok for now but for future PRs